third_party/libunwindstack/PeCoffEpilog.cpp

```diff
@@ -0,0 +1,240 @@
+      cs_x86_op operand0 = capstone_instruction_->detail->x86.operands[0];
+
+      // First operand is always a register for 'lea' instruction.
+      CHECK(operand0.type == X86_OP_REG);
+
+      x86_reg reg = operand0.reg;
+      if (reg != X86_REG_RSP) {
+        // The register that we set must be rsp, o/w we are not in the epilog.
+        return false;
+      }
+      cs_x86_op operand1 = capstone_instruction_->detail->x86.operands[1];
+
+      // Second operand is always a mem operand for 'lea' instructions.
+      CHECK(operand1.type == X86_OP_MEM);
+
+      // TODO: Not sure if this is really illegal.
+      CHECK(operand1.mem.segment == X86_REG_INVALID);
+
+      if (operand1.mem.index != X86_REG_INVALID) {
+        // Only instructions of the form lea rsp, constant[frame_pointer_register] are legal. This
+        // excludes using an index register.
+        return false;
+      }
+
+      x86_reg base_reg = operand1.mem.base;
+      uint16_t unwindstack_base_reg = MapCapstoneToUnwindstackRegister(base_reg);
+
+      uint64_t effective_address = (*updated_regs)[unwindstack_base_reg] + operand1.mem.disp;
+      updated_regs->set_sp(effective_address);
+    } else if (is_first_iteration && capstone_instruction_->id == X86_INS_ADD) {
+      CHECK(capstone_instruction_->detail->x86.op_count == 2);
+      cs_x86_op operand0 = capstone_instruction_->detail->x86.operands[0];
+      cs_x86_op operand1 = capstone_instruction_->detail->x86.operands[1];
+      if (operand0.type != X86_OP_REG || operand1.type != X86_OP_IMM) {
+        // The 'add' instruction must be adding an immediate value to a register, o/w
+        // we are not in the epilog.
+        return false;
+      }
+      x86_reg reg = operand0.reg;
+      if (reg != X86_REG_RSP) {
+        // The register that we add to must be rsp, o/w we are not in the epilog.
+        return false;
+      }
+      int64_t immediate_value = operand1.imm;
+      if (immediate_value < 0) {
+        // The immediate value represents the stack allocation size, so it must be non-negative.
+        return false;
+      }
+      updated_regs->set_sp(updated_regs->sp() + static_cast<uint64_t>(immediate_value));
+    } else if (capstone_instruction_->id == X86_INS_POP) {
+      CHECK(capstone_instruction_->detail->x86.op_count == 1);
+      cs_x86_op operand = capstone_instruction_->detail->x86.operands[0];
+      CHECK(operand.type == X86_OP_REG);
+      x86_reg reg = operand.reg;
+      uint16_t unwindstack_reg = MapCapstoneToUnwindstackRegister(reg);
+
+      uint64_t value;
+      if (!process_memory->Read64(updated_regs->sp(), &value)) {
+        last_error_.code = ERROR_MEMORY_INVALID;
+        last_error_.address = updated_regs->sp();
+        return false;
+      }
+
+      updated_regs->set_sp(updated_regs->sp() + sizeof(uint64_t));
+      (*updated_regs)[unwindstack_reg] = value;
+    } else if (capstone_instruction_->id == X86_INS_RET ||
```

Comment:
Not sure if this is better, but you could also check for the instruction `group` containing `X86_GRP_RET` (e.g. see `FunctionFramePointerValidator::IsRetOrJumpInstruction`)

Reply:
I'm not 100% sure, would like to defer once I've had the chance to run more experiments.

Reply:
Yeah, fine for me. And we can defer if that feels more useful. I just wanted to make sure to not miss something.